kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/BigQueryWriter.java

```diff
@@ -145,6 +163,35 @@ public void writeRows(PartitionedTableId table,
         mostRecentException);
   }
 
+  /**
+   * Decide whether the failure is a partial failure or complete failure
+   * @param numOfFailedRows Number of rows that failed
+   * @param numOfAllRows Number of rows to insert
+   * @return isPartialFailure
+   */
+  private boolean isPartialFailure(int numOfFailedRows, int numOfAllRows) {
+    return numOfFailedRows == numOfAllRows;
+  }
+
+  /**
+   * Filter out succeed rows, and return a list of failed rows.
+   * @param rows The rows to write.
+   * @param failRowsIndex A set of failed row index
+   * @return A list of failed rows
+   */
+  private List<InsertAllRequest.RowToInsert> getFailedRows(List<InsertAllRequest.RowToInsert> rows,
+                                                           Set<Long> failRowsIndex) {
+    logger.info("{} rows succeeded, {} rows failed", rows.size() - failRowsIndex.size(), failRowsIndex.size());
+    List<InsertAllRequest.RowToInsert> failRows = new ArrayList<>();
+    for (int index = 0; index < rows.size(); index++) {
+      if (failRowsIndex.contains((long)index)) {
+        logger.info("Adding [row index %d] for retry", index);
```

Comment:
This sounds like it's going to be REALLY verbose. This seems like potentially useful data, but maybe it should be at the debug level instead of info?

Also, this method is just filtering the rows. We don't know what the caller will use that data for. Instead of saying that we're going to retry, maybe just say the row with that index failed?

Reply:
absolutely